### PR TITLE
WIP: xtrabackup implementation + allow_master

### LINKED ISSUE #4557: Vtctl should have an explicit mode that allows you to force Master to Backup
#### Feature Description
If a user ever is in a situation where there are no replica's it is currently impossible to take a backup without manually stopping the vttablet. There is a command called ChangeSlaveType, but it will not allow you to change from a MASTER to a REPLICA.  There should be an explicit flag on ChangeSlaveType that allows a user to change the tablet type without restart vttablet. This will then allow a user to set the tablet to REPLICA take a backup then set it back to MASTER without having to stop the tablet.

Alternatively the Backup command could have a flag ( perhaps `allow_emergency_backup_on_master` ) that will allow it to execute on the master  

#### Use Case(s)
If binlogs have been cycles and not suitable backup exists to recover from

### PR DESCRIPTION
**Pull request by deepthi**
Refactor backup/restore to abstract out the backup engine.
Current logic where we do all the work is now called `builtin`.
Initial support for `xtrabackup` backup and restore. Tests TBI.
Add allow_master command line option to vtctl Backup and BackupShard commands.

All feedback and suggestions on code structure, coupling etc. are welcome.

cc @derekperkins @dkhenry @demmer @rafael

### CODE REVIEW

File: go/vt/vtctl/backup.go [hunk 1]
```diff
@@ -61,6 +61,8 @@ func init() {
 
 func commandBackup(ctx context.Context, wr *wrangler.Wrangler, subFlags *flag.FlagSet, args []string) error {
 	concurrency := subFlags.Int("concurrency", 4, "Specifies the number of compression/checksum jobs to run simultaneously")
+	allowMaster := subFlags.Bool("allow_master", false, "Allows backups to be taken on master. Warning!! If you are using the builtin backup engine, this will shutdown your master mysql for as long as it takes to create a backup ")
```
Comment (acharis) [inline go/vt/vtctl/backup.go]: can we make it so that allow_master is a no-op if it would be harmful?

Comment (deepthi) [reply go/vt/vtctl/backup.go]: how would we define "harmful"? If there is a way to define that we can look into implementing it.

The idea behind the proposal in the original issue (#4557) was that the user is explicitly overriding the constraint that dis-allows backups from master.

Comment (deepthi) [reply go/vt/vtctl/backup.go]: @acharis One check I implemented is that even if allow_master is specified with a BackupTablet command, we will first try to do a backup using one of the other types of vttablet(REPLICA/RDONLY/SPARE). Only if none of those is available we will use MASTER. I know this isn't what you asked, but FYI.

**Comment by rafael** [comment]
Could we rebase this branch with: https://github.com/vitessio/vitess/pull/4653 please? 

So we don't see diffs from that PR in here.

**Comment by deepthi** [comment]
Moved the xtrabackup implementation to its own PR: #4695

**Comment by deepthi** [comment]
This PR is superseded by #4653 #4695 and #4699

**Pull request closed without merging**